### PR TITLE
End of a hard task

### PR DESCRIPTION
**Pull request by MatviiGD**

### CODE REVIEW

File: src/lessonFour/Algorithm.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package lessonFour;
+
+abstract class Algorithm {
```
Comment (rzherebgrid) [inline src/lessonFour/Algorithm.java]: public

File: src/lessonFour/Algorithm.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+
+    abstract String encrypt(String message, int key);
```
Comment (rzherebgrid) [inline src/lessonFour/Algorithm.java]: public

File: src/lessonFour/Enigma.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                System.err.println("unknown mode");
+                System.exit(1);
+        }
+    }
+
+
+    // вивести результати шифрування /дешифрування в STDERR of "-out"
+
+    private static void output(String message, String outPath) {
+        if (outPath.equals("")) {
```
Comment (rzherebgrid) [inline src/lessonFour/Enigma.java]: It's much better to have "".equals(outPath)

File: src/lessonFour/Enigma.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                System.err.println("can't write to " + outPath);
+                System.exit(1);
+            }
+        }
+    }
+
+
+    //  прочитати дані з файлу "-in" якщо він був а  "-data"  не було дано
+    private void readData() {
+        if (data.equals("")) {  // prefer "-data" over "-in"
```
Comment (rzherebgrid) [inline src/lessonFour/Enigma.java]: Same here

**Review by rzherebgrid — CHANGES_REQUESTED** [review]

**Review by rzherebgrid — APPROVED** [review]

**Comment by MatviiGD** [comment]
thank you

**Pull request merged**